Source/MMParser.m

```diff
@@ -710,6 +726,345 @@ - (MMElement *)_parseHorizontalRuleWithScanner:(MMScanner *)scanner
     return element;
 }
 
+
+#pragma mark - Checklist
+
+/**
+ Check the following markers:
+ 
+ - [ ]
+ - [x]
+ - [X]
+ + [ ]
+ * [ ]
+ 1. [ ]
+ 
+ @param scanner MMScanner
+ @return ture if the check list marker exist, false otherwise.
+ */
+- (BOOL)_parseChecklistMarkerWithScanner:(MMScanner *)scanner listType:(MMElementType*)type expectedListType:(MMElementType)expectedListType checked:(nullable BOOL*)checked
+{
+    // checking
+    if(expectedListType == MMElementTypeNone || expectedListType == MMElementTypeBulletedChecklist)
+    {
+        [scanner beginTransaction];
+        unichar nextChar = scanner.nextCharacter;
+        if (nextChar == '*' || nextChar == '-' || nextChar == '+')
+        {
+            [scanner advance];
+            if (scanner.nextCharacter == ' ')
+            {
+                [scanner advance];
+                if (scanner.nextCharacter == '[')
+                {
+                    [scanner advance];
+                    nextChar = scanner.nextCharacter;
+                    if (nextChar == ' ' || nextChar == 'X' || nextChar == 'x')
+                    {
+                        BOOL foundx = nextChar == 'X' || nextChar == 'x';
+                        [scanner advance];
+                        if (scanner.nextCharacter == ']')
+                        {
+                            [scanner advance];
+                            if (scanner.nextCharacter == ' ')
+                            {
```

Comment:
Thanks for the code review.

After some consideration, I think the current structure is easer to read and understand. 
Plus, _parseListMarkerWithScanner is not strictly using early returns style. =)